rust/main/chains/hyperlane-sealevel/src/transaction.rs

```diff
@@ -0,0 +1,136 @@
+use solana_transaction_status::{
+    EncodedTransaction, EncodedTransactionWithStatusMeta, UiCompiledInstruction, UiInstruction,
+    UiMessage,
+};
+
+use hyperlane_core::H512;
+use hyperlane_sealevel_mailbox::instruction::Instruction;
+
+use crate::utils::{decode_h512, from_base58};
+
+/// This function searches for a transaction which dispatches Hyperlane message and returns
+/// list of hashes of such transactions.
+///
+/// This function takes the mailbox program identifier and the identifier for PDA for storing
+/// a dispatched message and searches a message dispatch transaction in a list of transaction.
+/// The list of transaction is usually comes from a block. The function returns list of hashes
+/// of such transactions.
+///
+/// The transaction will be searched with the following criteria:
+///     1. Transaction contains Mailbox program id in the list of accounts.
+///     2. Transaction contains dispatched message PDA in the list of accounts.
+///     3. Transaction is performing message dispatch (OutboxDispatch).
+///
+/// * `mailbox_program_id` - Identifier of Mailbox program
+/// * `message_storage_pda_pubkey` - Identifier for dispatch message store PDA
+/// * `transactions` - List of transactions
+pub fn search_transactions(
+    mailbox_program_id: &Pubkey,
+    message_storage_pda_pubkey: &Pubkey,
+    transactions: Vec<EncodedTransactionWithStatusMeta>,
+) -> Vec<H512> {
+    transactions
+        .into_iter()
+        .filter_map(|tx| match (tx.transaction, tx.meta) {
+            // We support only transactions encoded as JSON
+            // We need none-empty metadata as well
+            (EncodedTransaction::Json(t), Some(m)) => Some((t, m)),
+            _ => None,
+        })
+        .filter_map(|(t, m)| {
+            let transaction_hash = match t.signatures.first() {
+                Some(h) => h,
+                None => return None, // if transaction is not signed, we continue the search
+            };
+
+            let transaction_hash = match decode_h512(&transaction_hash) {
+                Ok(h) => h,
+                Err(_) => return None, // if we cannot parse transaction hash, we continue the search
+            };
+
+            // We support only Raw messages initially
+            let message = match t.message {
+                UiMessage::Raw(m) => m,
+                _ => return None,
+            };
+
+            let instructions = match m.inner_instructions {
+                OptionSerializer::Some(ii) => ii
+                    .into_iter()
+                    .map(|iii| iii.instructions)
+                    .flatten()
+                    .flat_map(|ii| match ii {
+                        UiInstruction::Compiled(ci) => Some(ci),
+                        _ => None,
+                    })
+                    .collect::<Vec<UiCompiledInstruction>>(),
+                OptionSerializer::None | OptionSerializer::Skip => return None,
```

Comment:
Since we are going through transactions in a block, there can be transaction without `inner_instructions`, I assume. The transaction we are interested in should have `inner_instructions` (even if mailbox is called directly, it should call interchain payment master, I suppose). So, we just ignore transaction which cannot be dispatched message transaction here.